File: src/input/inputHandler.ts
/**
 * Input action types
 */
export type InputAction = 
  | 'move_left'
  | 'move_right'
  | 'create_ice_left'
  | 'create_ice_right'
  | 'jump'
  | 'pause'
  | 'restart'
  | 'menu'
  | 'select_level_1'
  | 'select_level_2'
  | 'select_level_3'
  | 'select_level_4'
  | 'select_level_5'
  | 'select_level_6'
  | 'start_game'
  | 'quit';

/**
 * Key binding configuration
 */
interface KeyBinding {
  key: string;
  action: InputAction;
  requireShift?: boolean;
  requireCtrl?: boolean;
  requireAlt?: boolean;
}

/**
 * Input handler for keyboard and mouse input
 */
export class InputHandler {
  private keys: Map<string, boolean> = new Map();
  private previousKeys: Map<string, boolean> = new Map();
  private keyBindings: KeyBinding[] = [];
  private actionCallbacks: Map<InputAction, () => void> = new Map();
  private isActive: boolean = true;

  constructor() {
    this.setupDefaultBindings();
    this.setupEventListeners();
  }

  /**
   * Setup default key bindings
   */
  private setupDefaultBindings(): void {
    this.keyBindings = [
      // Movement
      { key: 'j', action: 'move_left' },
      { key: 'l', action: 'move_right' },
      { key: 'ArrowLeft', action: 'move_left' },
      { key: 'ArrowRight', action: 'move_right' },
      
      // Ice creation
      { key: 'a', action: 'create_ice_left' },
      { key: 'd', action: 'create_ice_right' },
      
      // Jump (using space for jump)
      { key: ' ', action: 'jump' },
      
      // Game control
      { key: 'Escape', action: 'pause' },
      { key: 'r', action: 'restart' },
      { key: 'Tab', action: 'menu' },
      
      // Level selection
      { key: '1', action: 'select_level_1' },
      { key: '2', action: 'select_level_2' },
      { key: '3', action: 'select_level_3' },
      { key: '4', action: 'select_level_4' },
      { key: '5', action: 'select_level_5' },
      { key: '6', action: 'select_level_6' },
      
      // Game control
      { key: ' ', action: 'start_game' },
      
      // Quit (Alt+F4)
      { key: 'F4', action: 'quit', requireAlt: true },
    ];
  }

  /**
   * Setup event listeners
   */
  private setupEventListeners(): void {
    // Keyboard events
    window.addEventListener('keydown', this.handleKeyDown.bind(this));
    window.addEventListener('keyup', this.handleKeyUp.bind(this));
    
    // Prevent default behavior for game keys
    window.addEventListener('keydown', this.preventDefaultActions.bind(this));
    
    // Focus/blur handling
    window.addEventListener('blur', this.handleBlur.bind(this));
    window.addEventListener('focus', this.handleFocus.bind(this));
  }

  /**
   * Handle key down event
   */
  private handleKeyDown(event: KeyboardEvent): void {
    if (!this.isActive) return;

    const key = event.code || event.key;
    this.keys.set(key, true);

    // Check for bound actions
    this.checkKeyBindings(event);
  }

  /**
   * Handle key up event
   */
  private handleKeyUp(event: KeyboardEvent): void {
    const key = event.code || event.key;
    this.keys.set(key, false);
  }

  /**
   * Check for key bound actions
   */
  private checkKeyBindings(event: KeyboardEvent): void {
    for (const binding of this.keyBindings) {
      if (this.matchesBinding(binding, event)) {
        this.triggerAction(binding.action);
        break; // Only trigger first matching action
      }
    }
  }

  /**
   * Check if event matches key binding
   */
  private matchesBinding(binding: KeyBinding, event: KeyboardEvent): boolean {
    const key = event.code || event.key;
    
    // Check main key
    if (key !== binding.key && event.key !== binding.key) {
      return false;
    }

    // Check modifier keys
    if (binding.requireShift && !event.shiftKey) return false;
    if (binding.requireCtrl && !event.ctrlKey) return false;
    if (binding.requireAlt && !event.altKey) return false;

    return true;
  }

  /**
   * Trigger an action
   */
  private triggerAction(action: InputAction): void {
    const callback = this.actionCallbacks.get(action);
    if (callback) {
      callback();
    }
  }

  /**
   * Prevent default actions for game keys
   */
  private preventDefaultActions(event: KeyboardEvent): void {
    const preventedKeys = [
      ' ', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight',
      'Tab', 'Escape'
    ];

    const key = event.code || event.key;
    if (preventedKeys.includes(key) || preventedKeys.includes(event.key)) {
      event.preventDefault();
    }
  }

  /**
   * Handle window blur (lose focus)
   */
  private handleBlur(): void {
    // Clear all keys when window loses focus
    this.keys.clear();
  }

  /**
   * Handle window focus (gain focus)
   */
  private handleFocus(): void {
    // Input is reactivated when window gains focus
    this.isActive = true;
  }

  /**
   * Update input state (call this each frame)
   */
  update(): void {
    // Store previous key states
    this.previousKeys = new Map(this.keys);
  }

  /**
   * Check if key is currently pressed
   */
  isKeyPressed(key: string): boolean {
    return this.keys.get(key) || false;
  }

  /**
   * Check if key was just pressed this frame
   */
  isKeyJustPressed(key: string): boolean {
    const current = this.keys.get(key) || false;
    const previous = this.previousKeys.get(key) || false;
    return current && !previous;
  }

  /**
   * Check if key was just released this frame
   */
  isKeyJustReleased(key: string): boolean {
    const current = this.keys.get(key) || false;
    const previous = this.previousKeys.get(key) || false;
    return !current && previous;
  }

  /**
   * Check if any movement key is pressed
   */
  isMovementPressed(): boolean {
    return this.isKeyPressed('j') || 
           this.isKeyPressed('l') || 
           this.isKeyPressed('ArrowLeft') || 
           this.isKeyPressed('ArrowRight');
  }

  /**
   * Get movement direction (-1 for left, 1 for right, 0 for none)
   */
  getMovementDirection(): number {
    if (this.isKeyPressed('j') || this.isKeyPressed('ArrowLeft')) {
      return -1;
    }
    if (this.isKeyPressed('l') || this.isKeyPressed('ArrowRight')) {
      return 1;
    }
    return 0;
  }

  /**
   * Check if ice creation key is pressed
   */
  isIceCreationPressed(): 'left' | 'right' | null {
    if (this.isKeyPressed('a')) return 'left';
    if (this.isKeyPressed('d')) return 'right';
    return null;
  }

  /**
   * Bind callback to action
   */
  bindActionCallback(action: InputAction, callback: () => void): void {
    this.actionCallbacks.set(action, callback);
  }

  /**
   * Unbind action callback
   */
  unbindActionCallback(action: InputAction): void {
    this.actionCallbacks.delete(action);
  }

  /**
   * Add custom key binding
   */
  addKeyBinding(binding: KeyBinding): void {
    this.keyBindings.push(binding);
  }

  /**
   * Remove key binding
   */
  removeKeyBinding(action: InputAction): void {
    this.keyBindings = this.keyBindings.filter(binding => binding.action !== action);
  }

  /**
   * Set input active state
   */
  setActive(active: boolean): void {
    this.isActive = active;
    if (!active) {
      this.keys.clear();
    }
  }

  /**
   * Get input active state
   */
  isInputActive(): boolean {
    return this.isActive;
  }

  /**
   * Clear all input state
   */
  clear(): void {
    this.keys.clear();
    this.previousKeys.clear();
  }

  /**
   * Get currently pressed keys (for debugging)
   */
  getPressedKeys(): string[] {
    return Array.from(this.keys.entries())
      .filter(([, pressed]) => pressed)
      .map(([key]) => key);
  }

  /**
   * Destroy input handler (remove event listeners)
   */
  destroy(): void {
    window.removeEventListener('keydown', this.handleKeyDown);
    window.removeEventListener('keyup', this.handleKeyUp);
    window.removeEventListener('keydown', this.preventDefaultActions);
    window.removeEventListener('blur', this.handleBlur);
    window.removeEventListener('focus', this.handleFocus);
    
    this.keys.clear();
    this.previousKeys.clear();
    this.actionCallbacks.clear();
  }
}

File: src/input/input_handler.py
# Input Handler for Game Controls

from typing import Dict, Set, Callable, Optional

try:
    import pygame
    PYGAME_AVAILABLE = True
except ImportError:
    PYGAME_AVAILABLE = False
    pygame = None

from src.game.constants import *


class InputHandler:
    """Handles all game input processing"""
    
    def __init__(self):
        self.keys_pressed: Set[int] = set()
        self.keys_just_pressed: Set[int] = set()
        self.keys_just_released: Set[int] = set()
        self.key_bindings: Dict[str, int] = {}
        self.action_callbacks: Dict[str, Callable] = {}
        
        # Initialize default key bindings
        self._setup_default_bindings()
    
    def _setup_default_bindings(self):
        """Setup default key bindings"""
        if not PYGAME_AVAILABLE:
            return
            
        self.key_bindings = {
            'move_left': pygame.K_j,
            'move_right': pygame.K_l,
            'ice_left': pygame.K_a,
            'ice_right': pygame.K_d,
            'alt_left': pygame.K_LEFT,
            'alt_right': pygame.K_RIGHT,
            'pause': pygame.K_ESCAPE,
            'restart': pygame.K_r,
            'menu': pygame.K_TAB,
        }
    
    def update(self):
        """Update input state (call once per frame)"""
        if not PYGAME_AVAILABLE:
            return
            
        # Get current keyboard state
        current_keys = pygame.key.get_pressed()
        current_key_set = set(i for i, pressed in enumerate(current_keys) if pressed)
        
        # Calculate just pressed and just released keys
        self.keys_just_pressed = current_key_set - self.keys_pressed
        self.keys_just_released = self.keys_pressed - current_key_set
        self.keys_pressed = current_key_set
        
        # Process action callbacks
        self._process_actions()
    
    def _process_actions(self):
        """Process action callbacks for just pressed keys"""
        for action, key_code in self.key_bindings.items():
            if self.is_key_just_pressed(key_code):
                callback = self.action_callbacks.get(action)
                if callback:
                    callback()
    
    def is_key_pressed(self, key: int) -> bool:
        """Check if a key is currently pressed"""
        return key in self.keys_pressed
    
    def is_key_just_pressed(self, key: int) -> bool:
        """Check if a key was just pressed this frame"""
        return key in self.keys_just_pressed
    
    def is_key_just_released(self, key: int) -> bool:
        """Check if a key was just released this frame"""
        return key in self.keys_just_released
    
    def is_action_pressed(self, action: str) -> bool:
        """Check if an action key is currently pressed"""
        key = self.key_bindings.get(action)
        if key:
            return self.is_key_pressed(key)
        return False
    
    def is_action_just_pressed(self, action: str) -> bool:
        """Check if an action key was just pressed this frame"""
        key = self.key_bindings.get(action)
        if key:
            return self.is_key_just_pressed(key)
        return False
    
    def is_action_just_released(self, action: str) -> bool:
        """Check if an action key was just released this frame"""
        key = self.key_bindings.get(action)
        if key:
            return self.is_key_just_released(key)
        return False
    
    def bind_action(self, action: str, key: int):
        """Bind an action to a key"""
        self.key_bindings[action] = key
    
    def bind_action_callback(self, action: str, callback: Callable):
        """Bind a callback function to an action"""
        self.action_callbacks[action] = callback
    
    def unbind_action_callback(self, action: str):
        """Remove callback for an action"""
        if action in self.action_callbacks:
            del self.action_callbacks[action]
    
    def get_key_for_action(self, action: str) -> Optional[int]:
        """Get the key code bound to an action"""
        return self.key_bindings.get(action)
    
    def get_action_name_for_key(self, key: int) -> Optional[str]:
        """Get the action name bound to a key"""
        for action, key_code in self.key_bindings.items():
            if key_code == key:
                return action
        return None
    
    def is_move_left_pressed(self) -> bool:
        """Check if move left is pressed (J or Left Arrow)"""
        return (self.is_action_pressed('move_left') or 
                self.is_action_pressed('alt_left'))
    
    def is_move_right_pressed(self) -> bool:
        """Check if move right is pressed (L or Right Arrow)"""
        return (self.is_action_pressed('move_right') or 
                self.is_action_pressed('alt_right'))
    
    def is_ice_left_pressed(self) -> bool:
        """Check if ice left is pressed (A key)"""
        return self.is_action_just_pressed('ice_left')
    
    def is_ice_right_pressed(self) -> bool:
        """Check if ice right is pressed (D key)"""
        return self.is_action_just_pressed('ice_right')
    
    def is_pause_pressed(self) -> bool:
        """Check if pause is pressed (ESC key)"""
        return self.is_action_just_pressed('pause')
    
    def is_restart_pressed(self) -> bool:
        """Check if restart is pressed (R key)"""
        return self.is_action_just_pressed('restart')
    
    def is_menu_pressed(self) -> bool:
        """Check if menu is pressed (TAB key)"""
        return self.is_action_just_pressed('menu')